### DISCUSSION TITLE
The relationship of generalizability of GNN and the dataset size

### DISCUSSION THREAD

**Discussion by YiKangOY** [pyg-team/pytorch_geometric#7251] [Q&A]
Hi all,

Now I have a graph-level regression task. I have a dataset of 39 graphs, now I use 34 to train and 5 to test.
The model I use is GIN and I use two layers. I use a sum pooling to get the embedding of the whole graph.

The result on the test graphs is really bad. I am wondering if this dataset is too small for a graph-level task so the model generalizes badly.

**Comment by rusty1s** [comment]
This is hard to tell, your dataset looks indeed quite small. Does the test performance actually increase over the number of epochs? What happens if you add some common regularization techniques?

**Reply by YiKangOY** [reply]
Hi, thanks for your reply.

I notice that the choice of test graphs will highly affect the test performance. The test performance increased at the first about 500-1000 epochs, and then become steady (I trained for 3000 epochs in total as at that point the performance on the training set do not further increase). So I think my model cannot generalize well on such a small dataset?

About the regularization method, now I used dropout(p=0.4) in the MLP after GNN for regression and weight_decay=2e-6 in Adam optimizer. The learning rate is 1e-3. Could you tell me more about the regularization techniques that I can use?

Here I attach a learning curve with a relatively good result with a mean MAPE of 21%.
![image](https://user-images.githubusercontent.com/63529290/235140123-31770227-98de-4634-a505-a1772603fbc8.png)

**Reply by YiKangOY** [reply]
![image](https://user-images.githubusercontent.com/63529290/235141229-40150e05-c855-41ae-9db0-c5e8167eb617.png)
Here is another curve, the train and test MAPE  are decreasing in the same way, but only for the first a few hundred epochs, then they became steady

**Reply by YiKangOY** [reply]
Here is a relatively bad result. Oh, I forget to say that these different curves are obtained by changing the graphs for testing.
![image](https://user-images.githubusercontent.com/63529290/235142758-7a5ac994-4e39-4fd0-b5b7-e3144509c1c1.png)

**Reply by rusty1s** [reply]
One thing you can also try is adding `BatchNorm` or `LayerNorm` to your model. I assume this might help to some extend in avoiding overfitting.

**Reply by YiKangOY** [reply]
I have used **Batchnorm** after each layer of GNN and I also used it in MLP. 
Should I use Batchnorm and Layernorm simultaneously or how should I choose between them?

**Reply by rusty1s** [reply]
Mh, okay. If you are already using normalization, then this is good. No need to apply batch norm and layer norm simultaneously.

**Reply by rusty1s** [reply]
One thing you can do is to confirm that your GNN actually does something reasonable. For example, an MLP should have worse evaluation performance. It might also be that your GIN model is too expressive, and you may have better luck with alternatives such as `SAGEConv`.

**Reply by YiKangOY** [reply]
Thanks, I will try other networks, but what do you mean by "For example, an MLP should have worse evaluation performance."? I just use MLP to regress for the target after GNN,

**Reply by rusty1s** [reply]
Meaning an MLP that replaces the GNN.
